module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,763 @@
+        let old_balance = balances.get(user_address, working_set).unwrap_or_default();
+        let total_balance = old_balance + bond_amount;
+        balances.set(user_address, &total_balance, working_set);
+
+        // Emit the bonding event
+        working_set.add_event(
+            event_key,
+            &format!("new_deposit: {bond_amount:?}. total_bond: {total_balance:?}"),
+        );
+
+        Ok(CallResponse::default())
+    }
+
+    /// Try to unbond the requested amount of coins with context.sender() as the beneficiary.
+    pub(crate) fn unbond_challenger(
+        &self,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        // Get the user's old balance.
+        if let Some(old_balance) = self.bonded_challengers.get(context.sender(), working_set) {
+            // Transfer the bond amount from the sender to the module's address.
+            // On failure, no state is changed
+            self.reward_sender(context, old_balance, working_set)?;
+
+            // Update our internal tracking of the total bonded amount for the sender.
+            self.bonded_challengers
+                .remove(context.sender(), working_set);
+
+            // Emit the unbonding event
+            working_set.add_event(
+                "unbonded_challenger",
+                &format!("amount_withdrawn: {old_balance:?}"),
+            );
+        }
+
+        Ok(CallResponse::default())
+    }
+
+    /// The attester starts the first phase of the two phase unbonding. We put the current max
+    /// finalized height with the attester address in the set of unbonding attesters iff the attester
+    /// is already present in the unbonding set
+    pub(crate) fn begin_unbond_attester(
+        &self,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // First get the bonded attester
+        if self
+            .bonded_attesters
+            .get(context.sender(), working_set)
+            .is_some()
+        {
+            // Then add the bonded attester to the unbonding set, with the current finalized height
+            let finalized_height = self
+                .light_client_finalized_height
+                .get(working_set)
+                .expect("Must be set at genesis");
+            self.unbonding_attesters
+                .set(context.sender(), &finalized_height, working_set);
+        }
+
+        Ok(CallResponse::default())
+    }
+
+    pub(crate) fn end_unbond_attester(
+        &self,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // We first have to ensure that the attester is bonded
+        if self
+            .bonded_attesters
+            .get(context.sender(), working_set)
+            .is_some()
```

Comment:
If we do remove the attester at the beginning of unbonding, we'll need to change this logic too

Reply:
Fixed with the last commit